service/history/workflow/mutable_state_impl.go

```diff
@@ -1157,7 +1182,11 @@ func (ms *MutableStateImpl) UpdateActivityProgress(
 	ai.LastHeartbeatDetails = request.Details
 	now := ms.timeSource.Now()
 	ai.LastHeartbeatUpdateTime = &now
+	if prev, existed := ms.updateActivityInfos[ai.ScheduledEventId]; existed {
```

Comment:
check pending map instead of the update map. update/delete map is cleared after each mutable state transaction. Plz double check other places.

Also the input `ai` is the current activity info.